### PR TITLE
Ensure default fund

### LINKED ISSUE #81: When first fund is assigned to the user it should be marked as default

### PR DESCRIPTION
**Pull request by Nuanda**
Fixes #81

### CODE REVIEW

File: app/models/atmosphere/user_fund.rb [hunk 1]
```diff
@@ -22,5 +22,37 @@ class UserFund < ActiveRecord::Base
...
+        end
+      end
+    end
+
+    def reallocate_default
+      ufs = UserFund.where(user_id: user_id)
+      if ufs.count > 0 && ufs.none?(&:default)
+        ufs.first.update_attributes(default: true)
+      end
+    end
```
Comment (mkasztelnik) [inline app/models/atmosphere/user_fund.rb]: This method can be simplified into:

```ruby
def reallocate_default
  uf = UserFund.find_by(user_id: user_id)
  uf.update_attributes(default: true) if default && uf
end
```

File: app/views/atmosphere/admin/funds/_user_fund.html.haml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      = user.login
+
+  - if user_fund.default
+    %span{ class: 'btn btn-xs btn-primary has_bottom_tooltip',
+    title: t('funds.set_default.to_false',
+             user: user_fund.user.full_name) }
+      D
+  - else
+    .btn-group
+      = form_for [:admin, user_fund] do |f|
```
Comment (mkasztelnik) [inline app/views/atmosphere/admin/funds/_user_fund.html.haml]: This have one small drawback - whole webpage is refreshed and selection for fund graph disappeared. This is very, very minor issue and if complicated don't have to be changed :smile:

File: spec/models/atmosphere/user_fund_spec.rb [hunk 3]
```diff
@@ -33,4 +33,75 @@
...
+
+    it 'makes first user_fund default' do
+      expect(user_fund.default).to be_truthy
+    end
+
+    it 'blocks setting the only fund as not default' do
+      expect(user_fund.default).to be_truthy
+      user_fund.default = false
+      user_fund.save
+      expect(user_fund.reload.default).to be_truthy
```
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: Maybe instead of reloading the object you can check for `user_fund` errors?

File: spec/models/atmosphere/user_fund_spec.rb [hunk 4]
```diff
@@ -33,4 +33,75 @@
...
+
+    it 'does not bother single default fund creation' do
+      pc = create(:poor_chap)
+      uf = create(:user_fund, user: pc, default: true)
+      expect(uf.reload.default).to be_truthy
+    end
+
+    it 'keeps only single default user fund' do
+      uf = create(:user_fund, user: user_fund.user, default: true)
+      expect(Atmosphere::UserFund.count).to eq 2
```
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: Why this is needed? Here you know exactly that `uf` and `user_fund` are in the database.

File: spec/models/atmosphere/user_fund_spec.rb [hunk 5]
```diff
@@ -33,4 +33,75 @@
...
+      expect(user_fund.reload.default).to be_falsey
+    end
+
+    it 'does not prematurely default to an invalid fund' do
+      begin
+        create(:user_fund,
+               fund: user_fund.fund,
+               user: user_fund.user,
+               default: true)
+      rescue ActiveRecord::RecordInvalid
```
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: This can be changed into following construction:

```ruby
expect do
  create(:user_fund, ...)
end.to raise_error(ActiveRecord::RecordInvalid)
```

or (maybe even better)

```ruby
expect do
  create(:user_fund, ...)
end.to change { Atmosphere::UserFund.count }.by 0
```

File: spec/models/atmosphere/user_fund_spec.rb [hunk 6]
```diff
@@ -33,4 +33,75 @@
...
+      expect(user_fund.reload.default).to be_truthy
+    end
+  end
+
+
+  describe '#reallocate_default' do
+
+    it 'does nothing when last fund is removed' do
+      user_fund.destroy
+      expect(Atmosphere::UserFund.count).to eq 0
```
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: I'm not sure if I understand the purpose of this test.

File: spec/models/atmosphere/user_fund_spec.rb [hunk 7]
```diff
@@ -33,4 +33,75 @@
...
+    end
+
+    it 'reallocates removed default' do
+      uf = create(:user_fund, user: user_fund.user)
+      uf2 = create(:user_fund, user: user_fund.user)
```
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: The is already spec checking if only one user fund can be set to default. As a conclusion maybe only one additional user fund (`uf`) can be created here (and the same situation in spec below) to check reallocation conditions?

**Comment by mkasztelnik** [comment]
I have only very small remarks/propositions for improvements, but the only thing I can write here is :+1:

 :clap: :clap:

File: app/models/atmosphere/user_fund.rb (hunk 1 above)
Comment (mkasztelnik) [inline app/models/atmosphere/user_fund.rb]: One additional improvement

```ruby
after_destroy :reallocate_default if :default

def reallocate_default
  uf = UserFund.find_by(user_id: user_id)
  uf && uf.update_attributes(default: true)
end
```

Comment (Nuanda) [inline app/models/atmosphere/user_fund.rb]: all right

File: app/views/atmosphere/admin/funds/_user_fund.html.haml (hunk 2 above)
Comment (Nuanda) [inline app/views/atmosphere/admin/funds/_user_fund.html.haml]: Yes, I know. Alternatives are quite complicated imho though.

File: spec/models/atmosphere/user_fund_spec.rb (hunk 3 above)
Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: There will be no errors since this is not regarding validation. Without reload most of those tests gave me false positives or false negatives. So I decided to be on the safe side so I reload them.

File: spec/models/atmosphere/user_fund_spec.rb (hunk 4 above)
Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: Sure. Unless the user factory (and you need a user there) creates two different funds just for the fun of it! So I end up spending a better part of an hour figuring these things out. Then I switched to my better :poor_chap factory but I decided to those leave "theoretically stupid" (but, in fact, quite useful) expectations in the code.

File: spec/models/atmosphere/user_fund_spec.rb (hunk 5 above)
Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: Yes, but then @mkasztelnik will tell me this spec tests for two different things and it is not acceptable by our standards ;).

File: spec/models/atmosphere/user_fund_spec.rb (hunk 6 above)
Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: Very useful. Tests for boundary cases - something we should test more. Perhaps, actually, those are the only spec tests worth doing (since many errors are because, for instance, callbacks assuming there is at least one object in the DB and failing with exceptions where there are none). In such a case this test should end with an exception and that will be easier to notice by a developer.

File: spec/models/atmosphere/user_fund_spec.rb (hunk 7 above)
Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: No, since we have different callbacks called here. So no - this was not tested before (saving/creating/updating was).

**Comment by Nuanda** [comment]
Marek, If you want to commit the changes regarding the callback syntax, and all tests pass, then go ahead.

File: spec/models/atmosphere/user_fund_spec.rb (hunk 4 above)
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: Ok I understand. I also discovered this issue that some of the factories do to much. We need to take a look at them in the feature

File: spec/models/atmosphere/user_fund_spec.rb (hunk 5 above)
Comment (mkasztelnik) [inline spec/models/atmosphere/user_fund_spec.rb]: Eeee, I'm not that bad :smiling_imp:

Comment (Nuanda) [inline spec/models/atmosphere/user_fund_spec.rb]: Just kidding. Go ahead and add the raise expectation, if you have a second.

**Comment by Nuanda** [comment]
I would ignore the self.default complaints of the houndini this time. First of all, they are acutally not all redundant, I think. And single 'default' is such a generic name, I decided to place the exact self. reference (what I usually _not_ do) to make it easier to read and understand.

**Comment by mkasztelnik** [comment]
@Nuanda as far as I know `self` is required only when you are setting value, for getters for sure `self` is redundant - but I understand your rationale :smile: 

:+1:

**Comment by Nuanda** [comment]
Thanks for commits.

**Comment by Nuanda** [comment]
@dice-cyfronet/atmo-dev-team another code review please

**Comment by nowakowski** [comment]
:+1:
